bitcoin/src/blockdata/script.rs

```diff
@@ -3,75 +3,606 @@
+    }
+}
 
-use crate::io;
-use core::convert::TryFrom;
-use core::{fmt, default::Default};
-use core::ops::Index;
-use bitcoin_internals::debug_from_display;
-#[cfg(feature = "bitcoinconsensus")]
-use bitcoin_internals::write_err;
+/// Iterator over bytes of a script
+pub struct Bytes<'a>(core::iter::Copied<core::slice::Iter<'a, u8>>);
 
-#[cfg(feature = "serde")] use serde;
+impl Iterator for Bytes<'_> {
+    type Item = u8;
 
-use crate::hash_types::{PubkeyHash, WPubkeyHash, ScriptHash, WScriptHash};
-use crate::blockdata::opcodes;
-use crate::blockdata::opcodes::all::*;
-use crate::consensus::{encode, Decodable, Encodable};
-use crate::hashes::{Hash, hex};
-use crate::policy::DUST_RELAY_TX_FEE;
-#[cfg(feature="bitcoinconsensus")] use bitcoinconsensus;
-#[cfg(feature="bitcoinconsensus")] use core::convert::From;
-use crate::OutPoint;
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next()
+    }
 
-use crate::crypto::key::PublicKey;
-use crate::address::WitnessVersion;
-use crate::taproot::{LeafVersion, TapBranchHash, TapLeafHash};
-use secp256k1::{Secp256k1, Verification, XOnlyPublicKey};
-use crate::crypto::schnorr::{TapTweak, TweakedPublicKey, UntweakedPublicKey};
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.0.size_hint()
+    }
 
-/// Bitcoin script.
-///
-/// A list of instructions in a simple, [Forth]-like, stack-based programming language
-/// that Bitcoin uses.
-///
-/// [Forth]: https://en.wikipedia.org/wiki/Forth_(programming_language)
-///
-/// See [Bitcoin Wiki: Script][wiki-script] for more information.
-///
-/// [wiki-script]: https://en.bitcoin.it/wiki/Script
-///
-/// ### Bitcoin Core References
-///
-/// * [CScript definition](https://github.com/bitcoin/bitcoin/blob/d492dc1cdaabdc52b0766bf4cba4bd73178325d0/src/script/script.h#L410)
-#[derive(Clone, Default, PartialOrd, Ord, PartialEq, Eq, Hash)]
-pub struct Script(Box<[u8]>);
+    #[inline]
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        self.0.nth(n)
+    }
+}
+
+impl ExactSizeIterator for Bytes<'_> {}
+impl core::iter::FusedIterator for Bytes<'_> {}
+
+macro_rules! delegate_index {
+    ($($type:ty),* $(,)?) => {
+        $(
+            /// Script subslicing operation - read [slicing safety](#slicing-safety)!
+            impl Index<$type> for Script {
+                type Output = Self;
+
+                #[inline]
+                fn index(&self, index: $type) -> &Self::Output {
+                    Self::from_bytes(&self.0[index])
+                }
+            }
+        )*
+    }
+}
 
-impl<I> Index<I> for Script
-where
-    [u8]: Index<I>,
-{
-    type Output = <[u8] as Index<I>>::Output;
+delegate_index!(Range<usize>, RangeFrom<usize>, RangeTo<usize>, RangeFull, RangeInclusive<usize>, RangeToInclusive<usize>);
+#[cfg(feature = "rust_v_1_53")]
+#[cfg_attr(docsrs, doc(cfg(feature = "rust_v_1_53")))]
+delegate_index!((Bound<usize>, Bound<usize>));
 
+impl AsRef<Script> for Script {
```

Comment:
What is the point of this impl? Is it just for uniformity (along with `as_mut`) so `ScriptBuf` and `Script` both have these implemented?

Reply:
It allows you to pass `&Script` to functions requiring `AsRef<Script>`. `std` cannot implement this as a blanket impl due to coherence issues (it does impl `AsRef<T> for &T` though) so types need to do it themselves. This is opposite of `Borrow` which has blanket `Borrow<T> for T` but not for references.

[`std` types have this impl too](https://doc.rust-lang.org/stable/std/path/struct.Path.html#impl-AsRef%3CPath%3E-for-Path) so it's also consistent with that.